include/dlaf/factorization/cholesky/impl.h

```diff
@@ -36,72 +36,81 @@ namespace factorization {
 namespace internal {
 
 namespace cholesky_l {
-template <class Executor, Device device, class T>
-void potrfDiagTile(Executor&& exec, hpx::future<matrix::Tile<T, device>> matrix_tile) {
-  hpx::dataflow(exec, matrix::unwrapExtendTiles(tile::potrf_o), blas::Uplo::Lower,
-                std::move(matrix_tile));
+template <Backend backend, class MatrixTileSender>
+void potrfDiagTile(MatrixTileSender&& matrix_tile) {
+  // TODO:
+  // - when_all(...) | algo(policy) | detach()?
+  // - algo(policy, when_all(...)) | detach?
+  // - detach(algo(policy, when_all(...)))?
+  tile::potrf(dlaf::internal::Policy<backend>(hpx::threads::thread_priority::normal),
+              dlaf::internal::whenAllLift(blas::Uplo::Lower,
+                                          std::forward<MatrixTileSender>(matrix_tile))) |
+      hpx::execution::experimental::detach();
 }
 
-template <class Executor, Device device, class T>
-void trsmPanelTile(Executor&& executor_hp, hpx::shared_future<matrix::Tile<const T, device>> kk_tile,
-                   hpx::future<matrix::Tile<T, device>> matrix_tile) {
-  hpx::dataflow(executor_hp, matrix::unwrapExtendTiles(tile::trsm_o), blas::Side::Right,
-                blas::Uplo::Lower, blas::Op::ConjTrans, blas::Diag::NonUnit, T(1.0), std::move(kk_tile),
-                std::move(matrix_tile));
+template <Backend backend, class T, class KKTileSender, class MatrixTileSender>
+void trsmPanelTile(KKTileSender&& kk_tile, MatrixTileSender&& matrix_tile) {
```

Comment:
Good poitns and this is also something I'm a bit worried about.

It's an inherent trade-off between allowing any sender that satisfies the sender concept to allow optimizations and type-erasing them to a (more) concrete type like `future<T>` (I'm planning to add something like an `any_sender<Ts...>` to HPX, but `future<T>` works just fine as a replacement in the meantime) and possibly paying for additional heap allocations.

Even with the the more generic sender version we could do a bit more than what's here. The senders can be constrained with the `sender_of` type trait/concept which gives some control over what they should send and the error messages. With SFINAE the error messages are of course not great. `static_assert`s might be a better option at the moment. Best would be C++20 concepts... I'll try something out but I fully agree that it's not clear where the trade-off should be. However, there's nothing stopping us from staying with concrete types like we have right now on master (whether they be `future`s or `any_sender`s) with the knowledge that we might have a small performance hit (it might be very small...).

As for the last question, it is indeed possible to get `T` from the sender, with some helper type traits.

Reply:
I had to go and check how much actually has changed.

About 200-250 of the added lines are documentation for the new overloads. The meat of the dispatch in `transform.h` is around 250 lines. This is clearly shorter than the executors that are there now (partly due to C++17 and partly due to a different strategy for CUDA streams/cuBLAS handles/cuSOLVER handles), but it's definitely not more complicated than the executors). The rest of the added lines are small helper utilities which will quite likely be easy to upstream. Things like `partial_transform` and the type list utilities are things that we have in HPX but which I didn't want to depend on at the moment. The rest of the changed lines are mostly moving the stream/handle pools to separate files and changing futures to senders and dataflows to sender algorithms.

Reply:
I've added some `static_assert`s to show roughly what we could do. If that looks acceptable I can apply them elsewhere as well. With concepts the checks would look roughly the same, but go in a `requires` clause. The main downside of constraining only using concepts is that e.g. this is not possible `template <typename T, Device device, sender_of<Tile<T, device>> TileSender>` f(TileSender&& t) { ... }`; `T` and `device` can't be inferred in this situation. As mentioned earlier, explicit, type-erased wrapper senders are still another possibility.

Reply:
Closing this as resolved. We decided that we can add constraints later if it turns out that using these is error prone. Ideally using concepts.